runners/spark/src/main/java/org/apache/beam/runners/spark/translation/MultiDoFnFunction.java

```diff
@@ -98,28 +110,108 @@ public MultiDoFnFunction(
 
     DoFnOutputManager outputManager = new DoFnOutputManager();
 
-    DoFnRunner<InputT, OutputT> doFnRunner =
+    SparkProcessContext.NoOpStepContext context;
+    InMemoryTimerInternals timerInternals = null;
+    // Now only implements the StatefulParDo in Batch mode.
+    if (stateful) {
+      Object key = null;
+      if (iter.hasNext()) {
+        WindowedValue<InputT> currentValue = iter.next();
+        key = ((KV) currentValue.getValue()).getKey();
+        iter = Iterators.concat(Iterators.singletonIterator(currentValue), iter);
+      }
+      final InMemoryStateInternals<?> stateInternals = InMemoryStateInternals.forKey(key);
+      timerInternals = new InMemoryTimerInternals();
+      final InMemoryTimerInternals finalTimerInternals = timerInternals;
```

Comment:
The `timerInternals/finalTimerInternals` is a little confusing to me. How about having just the one - `final InMemoryTimerInternals timeInternals` and setting `null` accordingly? just as you set the `NoOpStepContext`.